src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkWasmModule.java

```diff
@@ -0,0 +1,264 @@
+import net.starlark.java.annot.StarlarkBuiltin;
+import net.starlark.java.annot.StarlarkMethod;
+import net.starlark.java.eval.EvalException;
+import net.starlark.java.eval.Printer;
+import net.starlark.java.eval.Starlark;
+import net.starlark.java.eval.StarlarkValue;
+
+@Immutable
+@StarlarkBuiltin(
+    name = "wasm_module",
+    category = DocCategory.BUILTIN,
+    doc = "A WebAssembly module loaded by <code>repository_ctx.load_wasm()</code>.")
+final class StarlarkWasmModule implements StarlarkValue {
+  private final StarlarkPath path;
+  private final Object origPath;
+  private final WasmModule wasmModule;
+  private final String allocateFn;
+
+  public StarlarkWasmModule(
+      StarlarkPath path,
+      Object origPath,
+      byte[] moduleContent,
+      String allocateFn)
+      throws EvalException {
+    WasmModule wasmModule;
+    try {
+      wasmModule = com.dylibso.chicory.wasm.Parser.parse(moduleContent);
+    } catch (ChicoryException e) {
+      throw new EvalException(e);
+    }
+    validateModule(wasmModule, allocateFn);
+
+    this.path = path;
+    this.origPath = origPath;
+    this.wasmModule = wasmModule;
+    this.allocateFn = allocateFn;
+  }
+
+  @Override
+  public boolean isImmutable() {
+    return true;
+  }
+
+  @Override
+  public void repr(Printer printer) {
+    printer.append("<wasm_module path=");
+    printer.repr(origPath);
+    printer.append(">");
+  }
+
+  public StarlarkPath getPath() {
+    return path;
+  }
+
+  @StarlarkMethod(
+      name = "path",
+      structField = true,
+      doc = "The path this WebAssembly module was loaded from.")
+  public Object getOrigPath() {
+    return origPath;
+  }
+
+  public StarlarkWasmExecutionResult execute(
+      String functionName,
+      byte[] input,
+      Duration timeout,
+      long memLimitBytes)
+      throws EvalException, InterruptedException {
+    var memLimits = getMemLimits(memLimitBytes);
+    // Perform initialization and execution in a separate thread so it can be interrupted
+    // in case of timeout.
+    ExecutorService execService = Executors.newSingleThreadExecutor();
+    var execResultFuture = execService.submit(() -> {
+      return run(functionName, input, memLimits);
+    });
+    execService.shutdown();
+
+    StarlarkWasmExecutionResult result;
+    try {
+      result = execResultFuture.get(timeout.toMillis(), TimeUnit.MILLISECONDS);
+    } catch (TimeoutException _e) {
+      execService.shutdownNow();
+      while (!execService.isTerminated()) {
+        execService.awaitTermination(1, TimeUnit.SECONDS);
+      }
+      // Timeouts result in a `wasm_exec_result` with negative `return_code`.
+      result = new StarlarkWasmExecutionResult(-1, "");
+    } catch (ExecutionException e) {
+      // FIXME: This branch will be entered if out-of-memory occurs. Should such
+      // a condition be an exception, or act like a timeout and return a result
+      // with a negative `return_code`?
```

Comment:
As we want execution to be as isolated from the rest of Bazel as possible and the repo rule can show better error messages, I would vote for a negative return code.

Reply:
Done.